sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1118,14 +1133,46 @@ evaluated as normal during command graph execution.
 [source,c++]
 ----
 using namespace ext::oneapi::experimental;
-auto node = graph.add([&](sycl::handler& cgh){
+auto node = graph.add([&](device, sycl::handler& cgh){
 // Host code here is evaluated during the call to add()
 cgh.host_task([=](){
   // Code here is evaluated as part of executing the command graph node
 });
 });
 ----
 
+=== Multi-device Graph
+
+A user can compose a graph with nodes targeting different devices. This allows
+the benefits of defining an execution graph ahead of submission to be extended
+to multi-device platforms. Without this capability a user would have to submit
+individual single-device graphs and use events for dependencies, which is a
+usage model this extension is aiming to optimize. Automatic load balancing of
+commands across devices is not a problem this extension currently aims to solve,
+it is the responsibility of the user to decide the device each command will run
+on, not the SYCL runtime.
+
+A modifiable state `command_graph` contains nodes targeting specific devices,
+rather than being a device agnostic representation only tied to devices on
+finalization. This allows the implementation to process nodes which require
+device information when the command group function is evaluated. For example,
+a SYCL reduction implementation may desire the work-group/sub-group size, which
+is normally gathered by the runtime from the device associated with the queue.
+
+A drawback of this design is that it is less convenient for the use-case where
+a user would like to run the same graph on N devices. Rather than finalizing a
+single modifiable graph N times for N devices in this scenario, the user now has
+to record N modifiable graphs and then as finalize each of them. If this use
+case does become a usability issue, we could provide a specific API to support
+it. For example an update API for the modifiable `command_graph`, similar to
+parameters, so that the same modifiable graph can be updated to new devices then
+re-finalized.
```

Comment:
To better support this workflow, we mind consider supporting both sets of functions with and without specific device. The default device could be either set by the recording queue or through a property in explicit mode.

Reply:
I expect that a single device is a common use case and like the idea of having a default device.

Reply:
Agree that single device will be the most common use case. I have a slight conceptual concern in that the explicit API
is adding state to the graph to make the API less explicit. Mostly however my inclination to defer is just the pragmatic point of that if we want to say on Monday to the WG "please review this there are no more API changes planned" then we only really have a day and a half to add the default device". 

If we do add a default device, my main question is then whether a user should be able to change the default device, or if it's immutable on construction.

If a user can't change the default device, then a property to the existing graph constructor would be fine for setting the
default device. But it would be useful for a user to change the default device, then maybe a `set_default_device()`/`get_default_device()` pair of functions would be also be worth adding, as well as overloading the
current constructor with `command_graph(device)` too.

I don't think these are major issues, and I think we could work them out by the April 3rd WG even, but not sure if we could get agree->spec->review->merge it for Monday.

Reply:
Going to merge this PR as-is, so it's in for Mondays WG call but I've created Issue #109 to track this feature, and will bring it up to the WG on Monday for feedback. If it goes down well, then should be able to do a quick follow-up PR before the April 3rd call.